File: mobile/lib/models/food_item.dart
class FoodItem {
  final int id;
  final int providerId;
  final int? categoryId;
  final String name;
  final String slug;
  final String? description;
  final String? imageUrl;
  final double price;
  final double? discountPrice;
  final String unit;
  final int? preparationTimeMin;
  final bool isJain;
  final bool isVegan;
  final int spiceLevel; // 0-5
  final List<String> allergens;
  final bool isAvailable;
  final bool isFeatured;
  final double avgRating;
  final int totalOrders;

  const FoodItem({
    required this.id,
    required this.providerId,
    this.categoryId,
    required this.name,
    required this.slug,
    this.description,
    this.imageUrl,
    required this.price,
    this.discountPrice,
    this.unit = 'plate',
    this.preparationTimeMin,
    this.isJain = false,
    this.isVegan = false,
    this.spiceLevel = 0,
    this.allergens = const [],
    this.isAvailable = true,
    this.isFeatured = false,
    this.avgRating = 0.0,
    this.totalOrders = 0,
  });

  double get effectivePrice => discountPrice != null && discountPrice! > 0
      ? discountPrice!
      : price;

  bool get hasDiscount => discountPrice != null &&
      discountPrice! > 0 &&
      discountPrice! < price;

  factory FoodItem.fromJson(Map<String, dynamic> json) {
    return FoodItem(
      id: json['id'] as int,
      providerId: json['provider_id'] as int,
      categoryId: json['category_id'] as int?,
      name: json['name'] as String? ?? '',
      slug: json['slug'] as String? ?? '',
      description: json['description'] as String?,
      imageUrl: json['image_url'] as String?,
      price: (json['price'] as num? ?? 0).toDouble(),
      discountPrice: json['discount_price'] != null
          ? (json['discount_price'] as num).toDouble()
          : null,
      unit: json['unit'] as String? ?? 'plate',
      preparationTimeMin: json['preparation_time_min'] as int?,
      isJain: json['is_jain'] as bool? ?? false,
      isVegan: json['is_vegan'] as bool? ?? false,
      spiceLevel: json['spice_level'] as int? ?? 0,
      allergens: (json['allergens'] as List<dynamic>?)
              ?.map((e) => e as String)
              .toList() ??
          [],
      isAvailable: json['is_available'] as bool? ?? true,
      isFeatured: json['is_featured'] as bool? ?? false,
      avgRating: (json['avg_rating'] as num? ?? 0).toDouble(),
      totalOrders: json['total_orders'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'provider_id': providerId,
      'category_id': categoryId,
      'name': name,
      'slug': slug,
      'description': description,
      'image_url': imageUrl,
      'price': price,
      'discount_price': discountPrice,
      'unit': unit,
      'preparation_time_min': preparationTimeMin,
      'is_jain': isJain,
      'is_vegan': isVegan,
      'spice_level': spiceLevel,
      'allergens': allergens,
      'is_available': isAvailable,
      'is_featured': isFeatured,
      'avg_rating': avgRating,
      'total_orders': totalOrders,
    };
  }
}

class FoodCategory {
  final int id;
  final String name;
  final String? icon;
  final String? imageUrl;

  const FoodCategory({
    required this.id,
    required this.name,
    this.icon,
    this.imageUrl,
  });

  factory FoodCategory.fromJson(Map<String, dynamic> json) {
    return FoodCategory(
      id: json['id'] as int,
      name: json['name'] as String? ?? '',
      icon: json['icon'] as String?,
      imageUrl: json['image_url'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'icon': icon,
      'image_url': imageUrl,
    };
  }
}

File: app/Models/Food/FoodItem.php
<?php

namespace App\Models\Food;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class FoodItem extends Model
{
    use HasFactory;

    protected $fillable = [
        'provider_id',
        'category_id',
        'name',
        'slug',
        'description',
        'image_url',
        'price',
        'discount_price',
        'unit',
        'min_quantity',
        'max_quantity',
        'preparation_time_min',
        'is_jain',
        'is_vegan',
        'spice_level',
        'allergens',
        'ingredients',
        'is_available',
        'is_featured',
        'status',
        'price_range',
        'available_days',
        'available_from',
        'available_to',
        'total_orders',
        'avg_rating',
    ];

    protected $casts = [
        'price' => 'decimal:2',
        'discount_price' => 'decimal:2',
        'min_quantity' => 'integer',
        'max_quantity' => 'integer',
        'preparation_time_min' => 'integer',
        'is_jain' => 'boolean',
        'is_vegan' => 'boolean',
        'allergens' => 'array',
        'is_available' => 'boolean',
        'is_featured' => 'boolean',
        'available_days' => 'array',
        'total_orders' => 'integer',
        'avg_rating' => 'decimal:2',
    ];

    // ── Relationships ────────────────────────────────────────────────

    /**
     * Get the provider that offers this item.
     */
    public function provider(): BelongsTo
    {
        return $this->belongsTo(FoodProvider::class, 'provider_id');
    }

    /**
     * Get the category this item belongs to.
     */
    public function category(): BelongsTo
    {
        return $this->belongsTo(FoodCategory::class, 'category_id');
    }

    /**
     * Get the media (photos/videos) for this item.
     */
    public function media(): HasMany
    {
        return $this->hasMany(FoodItemMedia::class, 'food_item_id');
    }

    /**
     * Get the pricing tiers for this item.
     */
    public function pricingTiers(): HasMany
    {
        return $this->hasMany(FoodPricingTier::class, 'food_item_id');
    }

    /**
     * Get the order items that include this item.
     */
    public function orderItems(): HasMany
    {
        return $this->hasMany(FoodOrderItem::class, 'food_item_id');
    }

    /**
     * Get the reviews for this item.
     */
    public function reviews(): HasMany
    {
        return $this->hasMany(FoodReview::class, 'food_item_id');
    }

    /**
     * Get the cart entries for this item.
     */
    public function cartItems(): HasMany
    {
        return $this->hasMany(FoodCart::class, 'food_item_id');
    }

    /**
     * Get the wishlist entries for this item.
     */
    public function wishlists(): HasMany
    {
        return $this->hasMany(FoodWishlist::class, 'food_item_id');
    }

    // ── Scopes ───────────────────────────────────────────────────────

    public function scopeAvailable($query)
    {
        return $query->where('is_available', true);
    }

    public function scopeFeatured($query)
    {
        return $query->where('is_featured', true);
    }

    public function scopeJain($query)
    {
        return $query->where('is_jain', true);
    }

    public function scopeVegan($query)
    {
        return $query->where('is_vegan', true);
    }

    public function scopeBySpiceLevel($query, string $level)
    {
        return $query->where('spice_level', $level);
    }

    // ── Helpers ──────────────────────────────────────────────────────

    public function isAvailableToday(): bool
    {
        $today = strtolower(now()->format('D')); // mon, tue, ...

        if (!in_array($today, $this->available_days ?? [])) {
            return false;
        }

        $now = now()->format('H:i:s');

        if ($this->available_from && $now < $this->available_from) {
            return false;
        }

        if ($this->available_to && $now > $this->available_to) {
            return false;
        }

        return true;
    }

    public function effectivePrice(): string
    {
        return $this->discount_price ?? $this->price;
    }

    public function hasDiscount(): bool
    {
        return $this->discount_price !== null && $this->discount_price < $this->price;
    }
}
